File: legacy_android/app/src/main/java/com/example/data/remote/stripe/StripePaymentManager.kt
package com.example.data.remote.stripe

import android.content.Context
import android.content.SharedPreferences
import android.util.Log
import com.example.BuildConfig
import com.example.data.model.StripePaymentReceipt
import com.example.data.model.SubscriptionTier
import com.example.data.remote.supabase.SupabaseService
import com.squareup.moshi.Moshi
import com.squareup.moshi.Types
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.withContext
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import java.util.UUID
import java.util.concurrent.TimeUnit

enum class StripeCardBrand(val displayName: String) {
    VISA("Visa"),
    MASTERCARD("Mastercard"),
    AMEX("American Express"),
    DISCOVER("Discover"),
    UNKNOWN("Card")
}

data class StripeCardInput(
    val cardNumber: String = "",
    val expMonth: String = "",
    val expYear: String = "",
    val cvc: String = "",
    val cardholderName: String = "",
    val postalCode: String = ""
) {
    val brand: StripeCardBrand
        get() {
            val clean = cardNumber.replace(" ", "")
            return when {
                clean.startsWith("4") -> StripeCardBrand.VISA
                clean.startsWith("51") || clean.startsWith("52") || clean.startsWith("53") ||
                        clean.startsWith("54") || clean.startsWith("55") || clean.startsWith("22") -> StripeCardBrand.MASTERCARD
                clean.startsWith("34") || clean.startsWith("37") -> StripeCardBrand.AMEX
                clean.startsWith("6011") || clean.startsWith("65") -> StripeCardBrand.DISCOVER
                else -> StripeCardBrand.UNKNOWN
            }
        }

    val isValid: Boolean
        get() {
            val clean = cardNumber.replace(" ", "")
            val monthInt = expMonth.toIntOrNull() ?: 0
            val yearInt = expYear.toIntOrNull() ?: 0
            return clean.length in 15..16 &&
                    monthInt in 1..12 &&
                    yearInt >= 24 &&
                    cvc.length in 3..4 &&
                    cardholderName.isNotBlank()
        }
}

class StripePaymentManager(
    private val context: Context,
    private val supabaseService: SupabaseService
) {
    private val prefs: SharedPreferences = context.getSharedPreferences("noteflow_stripe_billing", Context.MODE_PRIVATE)
    private val moshi = Moshi.Builder().build()
    private val receiptListAdapter = moshi.adapter<List<StripePaymentReceipt>>(
        Types.newParameterizedType(List::class.java, StripePaymentReceipt::class.java)
    )

    val publishableKey: String = try {
        BuildConfig.STRIPE_PUBLISHABLE_KEY.ifBlank { "pk_test_51OExecutiveNoteFlowStripeKey94827593" }
    } catch (e: Throwable) {
        "pk_test_51OExecutiveNoteFlowStripeKey94827593"
    }

    private val _pastReceipts = MutableStateFlow<List<StripePaymentReceipt>>(emptyList())
    val pastReceipts: StateFlow<List<StripePaymentReceipt>> = _pastReceipts.asStateFlow()

    private val _activePaymentMethod = MutableStateFlow<String?>(null)
    val activePaymentMethod: StateFlow<String?> = _activePaymentMethod.asStateFlow()

    init {
        loadPastReceipts()
        _activePaymentMethod.value = prefs.getString("active_payment_method", "Visa ending in 4242")
    }

    private fun loadPastReceipts() {
        val json = prefs.getString("receipts_json", null)
        if (json != null) {
            try {
                val list = receiptListAdapter.fromJson(json) ?: emptyList()
                _pastReceipts.value = list
            } catch (e: Exception) {
                _pastReceipts.value = emptyList()
            }
        } else {
            // Initial sample receipt for executive preview
            val initial = listOf(
                StripePaymentReceipt(
                    paymentIntentId = "pi_3N9xFlowExecutive01",
                    customerEmail = "elena@enterprise.com",
                    amountPaidFormatted = "$180.00",
                    currency = "USD",
                    cardBrand = "Visa",
                    cardLast4 = "4242",
                    tierTitle = "Executive Pro",
                    billingCycle = "Billed annually",
                    timestamp = System.currentTimeMillis() - 86400000L * 14,
                    status = "succeeded",
                    receiptUrl = "https://pay.stripe.com/receipts/acct_noteflow_01"
                )
            )
            _pastReceipts.value = initial
            saveReceipts(initial)
        }
    }

    private fun saveReceipts(list: List<StripePaymentReceipt>) {
        try {
            val json = receiptListAdapter.toJson(list)
            prefs.edit().putString("receipts_json", json).apply()
        } catch (e: Exception) {
            Log.e("StripePaymentManager", "Error saving receipts: ${e.message}")
        }
    }

    /**
     * Process real/simulated Stripe Payment for Subscription
     */
    suspend fun processPayment(
        card: StripeCardInput,
        tier: SubscriptionTier,
        isAnnual: Boolean,
        customerEmail: String,
        userId: String
    ): Result<StripePaymentReceipt> = withContext(Dispatchers.IO) {
        try {
            // Artificial delay to reflect real Stripe 3D Secure / PaymentIntent verification
            delay(1200)

            val cleanNum = card.cardNumber.replace(" ", "")
            val last4 = if (cleanNum.length >= 4) cleanNum.takeLast(4) else "4242"
            val brand = card.brand.displayName

            val amountCents = tier.getPriceCents(isAnnual)
            val amountFormatted = String.format("$%.2f", amountCents / 100.0)
            val paymentIntentId = "pi_" + UUID.randomUUID().toString().replace("-", "").take(24)

            val receipt = StripePaymentReceipt(
                paymentIntentId = paymentIntentId,
                customerEmail = customerEmail,
                amountPaidFormatted = amountFormatted,
                currency = "USD",
                cardBrand = brand,
                cardLast4 = last4,
                tierTitle = tier.title,
                billingCycle = if (isAnnual) "Billed annually" else "Billed monthly",
                timestamp = System.currentTimeMillis(),
                status = "succeeded",
                receiptUrl = "https://dashboard.stripe.com/test/payments/$paymentIntentId"
            )

            // Save to local receipts list
            val updated = listOf(receipt) + _pastReceipts.value
            _pastReceipts.value = updated
            saveReceipts(updated)

            // Update active payment method
            val pmDesc = "$brand ending in $last4"
            _activePaymentMethod.value = pmDesc
            prefs.edit().putString("active_payment_method", pmDesc).apply()

            // Record to Supabase PostgreSQL Database table `subscriptions`
            supabaseService.recordSubscriptionToDatabase(
                userId = userId,
                tier = tier,
                stripeCustomerId = "cus_${userId.take(8)}",
                stripePaymentIntentId = paymentIntentId,
                amountFormatted = amountFormatted
            )

            Result.success(receipt)
        } catch (e: Exception) {
            Log.e("StripePaymentManager", "Payment processing failed: ${e.message}")
            Result.failure(e)
        }
    }

    /**
     * Format raw card digits with standard 4-digit spacing
     */
    fun formatCardNumber(input: String): String {
        val digits = input.filter { it.isDigit() }.take(16)
        val sb = StringBuilder()
        for (i in digits.indices) {
            if (i > 0 && i % 4 == 0) {
                sb.append(" ")
            }
            sb.append(digits[i])
        }
        return sb.toString()
    }

    /**
     * Format MM/YY expiry
     */
    fun formatExpiry(input: String): String {
        val digits = input.filter { it.isDigit() }.take(4)
        return when {
            digits.length <= 2 -> digits
            else -> digits.substring(0, 2) + "/" + digits.substring(2)
        }
    }
}

File: lib/services/stripe_payment_manager.dart
import 'dart:async';
import 'package:intl/intl.dart';
import 'package:uuid/uuid.dart';
import '../models/models.dart';
import '../data/local/preferences_service.dart';
import 'supabase_service.dart';

class StripePaymentManager {
  static final StripePaymentManager instance = StripePaymentManager._init();

  final String publishableKey = "pk_test_51OExecutiveNoteFlowStripeKey94827593";

  List<StripePaymentReceipt> _pastReceipts = [];
  List<StripePaymentReceipt> get pastReceipts => _pastReceipts;

  String _activePaymentMethod = "Visa ending in 4242";
  String get activePaymentMethod => _activePaymentMethod;

  final StreamController<List<StripePaymentReceipt>> _receiptsController = StreamController<List<StripePaymentReceipt>>.broadcast();
  Stream<List<StripePaymentReceipt>> get receiptsStream => _receiptsController.stream;

  final StreamController<String> _pmController = StreamController<String>.broadcast();
  Stream<String> get pmStream => _pmController.stream;

  StripePaymentManager._init() {
    loadBillingInfo();
  }

  Future<void> loadBillingInfo() async {
    _pastReceipts = await PreferencesService.instance.getPastReceipts();
    _activePaymentMethod = await PreferencesService.instance.getActivePaymentMethod() ?? "Visa ending in 4242";
    _receiptsController.add(_pastReceipts);
    _pmController.add(_activePaymentMethod);
  }

  Future<StripePaymentReceipt> processPayment({
    required StripeCardInput card,
    required SubscriptionTier tier,
    required bool isAnnual,
    required String customerEmail,
    required String userId,
  }) async {
    // Artificial delay to reflect real Stripe 3D Secure / PaymentIntent verification
    await Future.delayed(const Duration(milliseconds: 1200));

    final cleanNum = card.cardNumber.replaceAll(" ", "");
    final last4 = cleanNum.length >= 4 ? cleanNum.substring(cleanNum.length - 4) : "4242";
    final brand = card.brand.displayName;

    final amountCents = tier.getPriceCents(isAnnual);
    final amountFormatted = NumberFormat.currency(symbol: '\$').format(amountCents / 100.0);
    final paymentIntentId = "pi_${const Uuid().v4().replaceAll('-', '').substring(0, 24)}";

    final receipt = StripePaymentReceipt(
      paymentIntentId: paymentIntentId,
      customerEmail: customerEmail,
      amountPaidFormatted: amountFormatted,
      currency: "USD",
      cardBrand: brand,
      cardLast4: last4,
      tierTitle: tier.title,
      billingCycle: isAnnual ? "Billed annually" : "Billed monthly",
      timestamp: DateTime.now().millisecondsSinceEpoch,
      status: "succeeded",
      receiptUrl: "https://dashboard.stripe.com/test/payments/$paymentIntentId",
    );

    // Save to local receipts list
    _pastReceipts = [receipt, ..._pastReceipts];
    await PreferencesService.instance.savePastReceipts(_pastReceipts);
    _receiptsController.add(_pastReceipts);

    // Update active payment method
    final pmDesc = "$brand ending in $last4";
    _activePaymentMethod = pmDesc;
    await PreferencesService.instance.setActivePaymentMethod(pmDesc);
    _pmController.add(_activePaymentMethod);

    // Record to Supabase PostgreSQL table
    await SupabaseService.instance.recordSubscriptionToDatabase(
      userId: userId,
      tier: tier,
      stripeCustomerId: "cus_${userId.length > 8 ? userId.substring(0, 8) : userId}",
      stripePaymentIntentId: paymentIntentId,
      amountFormatted: amountFormatted,
    );

    return receipt;
  }

  static String formatCardNumber(String input) {
    final digits = input.replaceAll(RegExp(r'\D'), '');
    final limited = digits.length > 16 ? digits.substring(0, 16) : digits;
    final buffer = StringBuffer();
    for (int i = 0; i < limited.length; i++) {
      if (i > 0 && i % 4 == 0) {
        buffer.write(' ');
      }
      buffer.write(limited[i]);
    }
    return buffer.toString();
  }

  static String formatExpiry(String input) {
    final digits = input.replaceAll(RegExp(r'\D'), '');
    final limited = digits.length > 4 ? digits.substring(0, 4) : digits;
    if (limited.length <= 2) {
      return limited;
    } else {
      return '${limited.substring(0, 2)}/${limited.substring(2)}';
    }
  }

  void dispose() {
    _receiptsController.close();
    _pmController.close();
  }
}
